oclint-rules/include/oclint/AbstractASTVisitorRule.h

```diff
@@ -12,7 +11,50 @@ template<typename T>
 class AbstractASTVisitorRule : public AbstractASTRuleBase, protected clang::RecursiveASTVisitor<T>
 {
     friend class clang::RecursiveASTVisitor<T>;
+private:
+    virtual bool shouldTraverseFile(clang::SourceLocation startLocation)
+    {
+        if(startLocation.isInvalid())
+        {
+            return false;
+        }
+
+        clang::SourceManager *sourceManager = &_carrier->getSourceManager();
+
+        /* If the decl is in the current file, traverse it */
+        if (sourceManager->getMainFileID() == sourceManager->getFileID(startLocation))
+        {
+            return true;
+        }
+
+        /* If the user want to lint header files and decl is in the associated header file, traverse it */
+        if (!visitHeaders())
+        {
+            return false;
+        }
+
+        llvm::StringRef sourceFileName = sourceManager->getFilename(startLocation);
+        return isAnalyzeFileHeadFile(sourceFileName);
+    }
+
+    /* Check if decl is in the associated header file */
+    virtual bool isAnalyzeFileHeadFile(llvm::StringRef fileNameStrRef) {
```

Comment:
The changes in general are in very good shape, thanks for working on it. I also like its direction. From an Objective-C world, this is pretty much all it is needed to get this pull request merged. But the concern I still have for this method is its application in C and C++. Because although it is not that frequently seen, it is not uncommon to see `.c` and `.cpp` implementations have different header file names, so depending on name matching may not work all the time. Can you help think about this a little bit further, and see if there are better alternatives?

Reply:
This is a very good point but I'm not sure it's solvable in a general way. What would be needed is a way to mark one particular header file as _the_ header file of _this_ implementation file. I don't think there is a mechanism in clang to be able to tell. The only thing that I can think of is passing the name of the header file to check as part of the command line arguments.

Reply:
How about this? I think we could include a comment that turns on header visiting for specific headers. What I'm thinking is that in the same fashion that there is a `//!OCLINT` comment that turns off listing for a statement, we could add a `//OCLINT` comment that turns it _on_ for a given header.

Then you could do:

```
#include "foo.h"
#include "bar.h" //OCLINT
```
and when listing that file, oclint would only visit `bar.h`